File: section9/3-concurrency/concurrency.cpp
#include <thread>
#include <mutex>
#include <vector>
#include <iostream>

std::vector<int> make_large_vector() {
    std::vector<int> v;
    for (int i = 0; i < 5000; i++) {
        v.push_back(i - 300);
    }
    return v;
}

void update_account(int* balance, std::vector<int>* requests, std::mutex* requests_mtx) {
    while (true) {
        requests_mtx->lock();
        if (requests->empty()) {
            requests_mtx->unlock();
            return;
        }
        int current_request = requests->at(0);
        requests->erase(requests->begin());
        requests_mtx->unlock();

        *balance += current_request;
    }
}

int main() {
    int balance = 300;
    std::vector<int> requests = make_large_vector();
    std::mutex requests_mtx;

    std::vector<std::thread> workers;
    for (int i = 0; i < 6; i++) {
        workers.push_back(std::thread(update_account, &balance, &requests, &requests_mtx));
    }
    for (int i = 0; i < 6; i++) {
        workers.at(i).join();
    }

    std::cout << "Final balance = " << balance << std::endl;
    return 0;
}

File: section9/3-concurrency/concurrency.rs
use std::sync::{Arc, Mutex};
use std::thread;

fn make_large_vector() -> Vec<i32> {
    let mut v = vec![];
    for i in 0..5000 {
        v.push(i - 300);
    }
    v
}

fn update_account(balance: &Mutex<i32>) {
    for i in 0..5000 {
        let mut requests_guard = requests.lock().unwrap();
        if requests_guard.is_empty() {
            return;
        }
        let current_request = requests_guard.get(0).unwrap().clone();
        requests_guard.remove(0);
        drop(requests_guard);
        
        *balance += current_request;
    }
}

fn main() {
    let balance = Arc::new(Mutex::new(300));
    let requests = Arc::new(Mutex::new(make_large_vector()));

    let mut workers = vec![];
    for _ in 0..6 {
        let balance_ref = balance.clone();
        let requests_ref = requests.clone();
        workers.push(thread::spawn(move || update_account(&*balance_ref, &*requests_ref)));
    }

    for worker in workers {
        worker.join().unwrap();
    }

    println!("Final balance = {:?}", *balance.lock().unwrap());
}